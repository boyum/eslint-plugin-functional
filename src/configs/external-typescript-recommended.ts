import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import externalVanillaRecommended from "#eslint-plugin-functional/configs/external-vanilla-recommended";

const tsConfig = {
  "@typescript-eslint/prefer-readonly": "error",
  "@typescript-eslint/switch-exhaustiveness-check": "error",
} satisfies FlatConfig.Config["rules"];

export default {
  ...externalVanillaRecommended,
  ...tsConfig,
} satisfies FlatConfig.Config["rules"];
