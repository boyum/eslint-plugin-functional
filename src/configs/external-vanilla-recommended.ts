import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

export default {
  "prefer-const": "error",
  "no-param-reassign": "error",
  "no-var": "error",
} satisfies FlatConfig.Config["rules"];
