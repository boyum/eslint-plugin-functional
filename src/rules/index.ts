import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClassInheritance from "./no-class-inheritance";
import * as noClasses from "./no-classes";
import * as noConditionalStatements from "./no-conditional-statements";
import * as noExpressionStatements from "./no-expression-statements";
import * as noLet from "./no-let";
import * as noLoopStatements from "./no-loop-statements";
import * as noMixedTypes from "./no-mixed-types";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpressions from "./no-this-expressions";
import * as noThrowStatements from "./no-throw-statements";
import * as noTryStatements from "./no-try-statements";
import * as preferImmutableTypes from "./prefer-immutable-types";
import * as preferPropertySignatures from "./prefer-property-signatures";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferTacit from "./prefer-tacit";
import * as readonlyType from "./readonly-type";
import * as typeDeclarationImmutability from "./type-declaration-immutability";

/**
 * All of the custom rules.
 */
export const rules: Readonly<{
  [functionalParameters.name]: typeof functionalParameters.rule;
  [immutableData.name]: typeof immutableData.rule;
  [noClasses.name]: typeof noClasses.rule;
  [noClassInheritance.name]: typeof noClassInheritance.rule;
  [noConditionalStatements.name]: typeof noConditionalStatements.rule;
  [noExpressionStatements.name]: typeof noExpressionStatements.rule;
  [noLet.name]: typeof noLet.rule;
  [noLoopStatements.name]: typeof noLoopStatements.rule;
  [noMixedTypes.name]: typeof noMixedTypes.rule;
  [noPromiseReject.name]: typeof noPromiseReject.rule;
  [noReturnVoid.name]: typeof noReturnVoid.rule;
  [noThisExpressions.name]: typeof noThisExpressions.rule;
  [noThrowStatements.name]: typeof noThrowStatements.rule;
  [noTryStatements.name]: typeof noTryStatements.rule;
  [preferImmutableTypes.name]: typeof preferImmutableTypes.rule;
  [preferPropertySignatures.name]: typeof preferPropertySignatures.rule;
  [preferReadonlyTypes.name]: typeof preferReadonlyTypes.rule;
  [preferTacit.name]: typeof preferTacit.rule;
  [readonlyType.name]: typeof readonlyType.rule;
  [typeDeclarationImmutability.name]: typeof typeDeclarationImmutability.rule;
}> = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClasses.name]: noClasses.rule,
  [noClassInheritance.name]: noClassInheritance.rule,
  [noConditionalStatements.name]: noConditionalStatements.rule,
  [noExpressionStatements.name]: noExpressionStatements.rule,
  [noLet.name]: noLet.rule,
  [noLoopStatements.name]: noLoopStatements.rule,
  [noMixedTypes.name]: noMixedTypes.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpressions.name]: noThisExpressions.rule,
  [noThrowStatements.name]: noThrowStatements.rule,
  [noTryStatements.name]: noTryStatements.rule,
  [preferImmutableTypes.name]: preferImmutableTypes.rule,
  [preferPropertySignatures.name]: preferPropertySignatures.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferTacit.name]: preferTacit.rule,
  [readonlyType.name]: readonlyType.rule,
  [typeDeclarationImmutability.name]: typeDeclarationImmutability.rule,
};
