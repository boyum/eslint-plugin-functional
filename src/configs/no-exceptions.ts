import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Exceptions" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
