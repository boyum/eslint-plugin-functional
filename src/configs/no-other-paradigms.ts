import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Other Paradigms" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
