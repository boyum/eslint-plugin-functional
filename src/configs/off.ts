import type { FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#/rules";
import { ruleNameScope } from "#/utils/misc";

export default Object.fromEntries(
  Object.entries(rules).map(([name]) => [`${ruleNameScope}/${name}`, "off"]),
) satisfies FlatConfig.Config["rules"] as NonNullable<FlatConfig.Config["rules"]>;
