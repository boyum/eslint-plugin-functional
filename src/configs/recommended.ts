import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";
import { Immutability } from "is-immutable-type";

import { rules } from "#eslint-plugin-functional/rules";
import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noLet from "#eslint-plugin-functional/rules/no-let";
import * as noThisExpressions from "#eslint-plugin-functional/rules/no-this-expressions";
import * as noThrowStatements from "#eslint-plugin-functional/rules/no-throw-statements";
import * as noTryStatements from "#eslint-plugin-functional/rules/no-try-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";
import * as typeDeclarationImmutability from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { RuleEnforcementComparator } from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

const recommended = Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.recommended === "recommended" &&
        rule.meta.docs.category !== "Stylistic",
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: {
        ignoreLambdaExpression: true,
        ignoreIIFE: true,
        ignoreGettersAndSetters: true,
      },
    },
  ],
  [noConditionalStatements.fullName]: [
    "error",
    {
      allowReturningBranches: true,
    },
  ],
  [noLet.fullName]: [
    "error",
    {
      allowInForLoopInit: true,
    },
  ],
  [noThisExpressions.fullName]: "off",
  [noThrowStatements.fullName]: [
    "error",
    {
      allowInAsyncFunctions: true,
    },
  ],
  [noTryStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyDeep",
      },
    },
  ],
  [typeDeclarationImmutability.fullName]: [
    "error",
    {
      rules: [
        {
          identifiers: ["^I?Immutable.+"],
          immutability: Immutability.Immutable,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?ReadonlyDeep.+"],
          immutability: Immutability.ReadonlyDeep,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?Readonly.+"],
          immutability: Immutability.ReadonlyShallow,
          comparator: RuleEnforcementComparator.AtLeast,
          fixer: [
            {
              pattern: "^(Array|Map|Set)<(.+)>$",
              replace: "Readonly$1<$2>",
            },
            {
              pattern: "^(.+)$",
              replace: "Readonly<$1>",
            },
          ],
        },
        {
          identifiers: ["^I?Mutable.+"],
          immutability: Immutability.Mutable,
          comparator: RuleEnforcementComparator.AtMost,
          fixer: [
            {
              pattern: "^Readonly(Array|Map|Set)<(.+)>$",
              replace: "$1<$2>",
            },
            {
              pattern: "^Readonly<(.+)>$",
              replace: "$1",
            },
          ],
        },
      ],
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
