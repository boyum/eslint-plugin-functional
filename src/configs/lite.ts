import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as immutableData from "#eslint-plugin-functional/rules/immutable-data";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noExpressionStatements from "#eslint-plugin-functional/rules/no-expression-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";

import recommended from "./recommended";

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: false,
    },
  ],
  [immutableData.fullName]: ["error", { ignoreClasses: "fieldsOnly" }],
  [noConditionalStatements.fullName]: "off",
  [noExpressionStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyShallow",
      },
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
