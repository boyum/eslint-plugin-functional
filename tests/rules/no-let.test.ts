import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/no-let";

import { esLatestConfig } from "../utils/configs";

describe(name, () => {
  describe("javascript - es latest", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: esLatestConfig,
    });

    it("should report let declarations", () => {
      const invalidResult = invalid({
        code: dedent`
          let x;

          function foo() {
            let y;
            let z = 0;
          }
        `,
        errors: ["generic", "generic", "generic"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    describe("options", () => {
      describe("allowInFunctions", () => {
        it("should not report let declarations in function declarations", () => {
          valid({
            code: dedent`
              function foo() {
                let x;
                let y = 0;
              }
            `,
            options: [{ allowInFunctions: true }],
          });

          const invalidResult = invalid({
            code: dedent`
              let x;
              let y = 0;
            `,
            options: [{ allowInFunctions: true }],
            errors: ["generic", "generic"],
          });
          expect(invalidResult.messages).toMatchSnapshot();
        });

        it("should not report let declarations in arrow function declarations", () => {
          valid({
            code: dedent`
              const foo = () => {
                let x;
                let y = 0;
              };
            `,
            options: [{ allowInFunctions: true }],
          });

          const invalidResult = invalid({
            code: dedent`
              let x;
              let y = 0;
            `,
            options: [{ allowInFunctions: true }],
            errors: ["generic", "generic"],
          });
          expect(invalidResult.messages).toMatchSnapshot();
        });

        it("should not report let declarations in method declarations", () => {
          valid({
            code: dedent`
              class Foo {
                foo() {
                  let x;
                  let y = 0;
                }
              }
            `,
            options: [{ allowInFunctions: true }],
          });

          const invalidResult = invalid({
            code: dedent`
              let x;
              let y = 0;
            `,
            options: [{ allowInFunctions: true }],
            errors: ["generic", "generic"],
          });
          expect(invalidResult.messages).toMatchSnapshot();
        });
      });

      describe("allowInForLoopInit", () => {
        it("should not report let declarations in for loop init", () => {
          valid({
            code: `for (let x = 0; x < 1; x++);`,
            options: [{ allowInForLoopInit: true }],
          });

          const invalidResult = invalid({
            code: dedent`
              let x;
              let y = 0;
            `,
            options: [{ allowInFunctions: true }],
            errors: ["generic", "generic"],
          });
          expect(invalidResult.messages).toMatchSnapshot();
        });
      });

      describe("ignoreIdentifierPattern", () => {
        it("should not report let declarations with matching identifiers", () => {
          valid({
            code: dedent`
              let mutable;
              let mutableX
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: dedent`
              let mutable = 0;
              let mutableX = 0
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: `for (let mutableX = 0; x < 1; x++);`,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: `for (let mutableX in {});`,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: `for (let mutableX of []);`,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: dedent`
              function foo() {
                let mutableX;
                let mutableY = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: dedent`
              const foo = () => {
                let mutableX;
                let mutableY = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: dedent`
              class Foo {
                foo() {
                  let mutableX;
                  let mutableY = 0;
                }
              }
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          valid({
            code: dedent`
              let Mutable;
              let xMutable;
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: dedent`
              let Mutable = 0;
              let xMutable = 0;
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: `for (let xMutable = 0; xMutable < 1; xMutable++);`,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: `for (let xMutable in {});`,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: `for (let xMutable of []);`,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: dedent`
              function foo() {
                let xMutable;
                let yMutable = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: dedent`
              const foo = () => {
                let xMutable;
                let yMutable = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          valid({
            code: dedent`
              class Foo {
                foo() {
                  let xMutable;
                  let yMutable = 0;
                }
              }
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });
        });

        it("should report non ignored let declarations", () => {
          invalid({
            code: dedent`
              let immutable;
              let immutableX;
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: dedent`
              let immutable = 0;
              let immutableX = 0;
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: `for (let immutableX = 0; immutableX < 1; immutableX++);`,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: `for (let immutableX in {});`,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: `for (let immutableX of []);`,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: dedent`
              function foo() {
                let immutableX;
                let immutableY = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: dedent`
              const foo = () => {
                let immutableX;
                let immutableY = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: dedent`
              class Foo {
                foo() {
                  let immutableX;
                  let immutableY = 0;
                }
              }
            `,
            options: [{ ignoreIdentifierPattern: "^mutable" }],
          });

          invalid({
            code: dedent`
              let Immutable;
              let xImmutable;
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: dedent`
              let Immutable = 0;
              let xImmutable = 0;
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: `for (let xImmutable = 0; x < 1; x++);`,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: `for (let xImmutable in {});`,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: `for (let xImmutable of []);`,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: dedent`
              function foo() {
                let xImmutable;
                let yImmutable = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: dedent`
              const foo = () => {
                let xImmutable;
                let yImmutable = 0;
              }
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });

          invalid({
            code: dedent`
              class Foo {
                foo() {
                  let xImmutable;
                  let yImmutable = 0;
                }
              }
            `,
            options: [{ ignoreIdentifierPattern: "Mutable$" }],
          });
        });
      });
    });
  });
});
