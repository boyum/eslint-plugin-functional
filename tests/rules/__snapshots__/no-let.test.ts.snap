// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no-let > javascript - es latest > options > allowInForLoopInit > should not report let declarations in for loop init 1`] = `
[
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 1,
    "line": 1,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 11,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
]
`;

exports[`no-let > javascript - es latest > options > allowInFunctions > should not report let declarations in arrow function declarations 1`] = `
[
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 1,
    "line": 1,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 11,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
]
`;

exports[`no-let > javascript - es latest > options > allowInFunctions > should not report let declarations in function declarations 1`] = `
[
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 1,
    "line": 1,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 11,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
]
`;

exports[`no-let > javascript - es latest > options > allowInFunctions > should not report let declarations in method declarations 1`] = `
[
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 1,
    "line": 1,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 11,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
]
`;

exports[`no-let > javascript - es latest > should report let declarations 1`] = `
[
  {
    "column": 1,
    "endColumn": 7,
    "endLine": 1,
    "line": 1,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
  {
    "column": 3,
    "endColumn": 9,
    "endLine": 4,
    "line": 4,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
  {
    "column": 3,
    "endColumn": 13,
    "endLine": 5,
    "line": 5,
    "message": "Unexpected let, use const instead.",
    "messageId": "generic",
    "nodeType": "VariableDeclaration",
    "ruleId": "no-let",
    "severity": 2,
  },
]
`;
